### PR TITLE
Adds indexes on embedding table

### PR DESCRIPTION
**Pull request by fmeheust**
This pull request adds the possibility to create IVF indexes on the embedding column and function-based index on the metadata column. This is a first upload the documentation is still missing.

### CODE REVIEW

File: langchain4j-oracle/src/test/java/dev/langchain4j/store/embedding/oracle/CommonTestOperations.java
Comment (Michael-A-McMahon) [inline langchain4j-oracle/src/test/java/dev/langchain4j/store/embedding/oracle/CommonTestOperations.java]: We could call it "SYSDBA_DATA_SOURCE" to reflect the difference from the normal DATA_SOURCE. We could also consider if we need to be pooling sysdba connections; If not then we could just have a CommonTestOperations.getSysDbaConnection() method.

File: langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/OracleEmbeddingStore.java
Comment (Michael-A-McMahon) [inline langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/OracleEmbeddingStore.java]: Wondering why the argument type is `IndexBuilder` and not `Index`

File: langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/OracleEmbeddingStore.java
Comment (Michael-A-McMahon) [inline langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/OracleEmbeddingStore.java]: We should not acquire a Connection unless we know we gonna use it. We should check if any index creation is actually required first.

File: langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/Index.java
Comment (Michael-A-McMahon) [inline langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/Index.java]: It seems like these public methods could be package-private instead.
We could even go further with the encapsulation, and replace these methods with a `create(DataSource, EmbeddingTable)` method. That could be nice as subclasses have more flexibility in how they create and drop objects. For instance, they might execute multiple statements. Or maybe we don't need it. Hard to guess what the future holds...

File: langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/OracleEmbeddingStore.java
Comment (fmeheust) [reply langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/OracleEmbeddingStore.java]: So that the build method can be called while constructing the store and pass the embedding table as parameter. Another option would be to leave placeholders on the create and drop statements and fill them when the store is being created. I will see what I can do.

**Comment by fmeheust** [comment]
I have incorporated the suggested changes. Can you please review again?

**Comment by Michael-A-McMahon** [comment]
> I have incorporated the suggested changes. Can you please review again?

Sorry, I couldn't get to this today. I'll try to review before Monday.

File: langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/IndexBuilder.java
Comment (Michael-A-McMahon) [inline langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/IndexBuilder.java:77]: What if we let the database raise an error in this case?
This way the user would know that their name is too long, and they can change it to a shorter one. My thought is that a user might expect the index to have the exact name they configured. If we use a different name, then the user won't know what to look for when using the index in another SQL command, like a DROP or a REBUILD.
Also, if a future database version supports longer names, then we won't be getting in the way of that.

File: langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/JSONIndexBuilder.java
Comment (Michael-A-McMahon) [inline langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/JSONIndexBuilder.java:98]: We could move these validations to the method where the value is set. This way, the user gets a stacktrace leading from the code where the value originates.

File: langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/OracleEmbeddingStore.java
Comment (Michael-A-McMahon) [inline langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/OracleEmbeddingStore.java]: Typo, should be past-tense: "... is created ..."

**Review by Michael-A-McMahon — APPROVED** [review]
I'm really happy with how this turned out. Thanks for the massive effort here; I never knew indexes were this complex!

Would be good to look over all docs for typos before merging.

Also consider the comment on name truncation. I can see the benefit of not doing this (no error for the user), and I can see the benefit of having the error too. I'll leave it up to your judgement.

File: langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/IndexBuilder.java
Comment (fmeheust) [reply langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/IndexBuilder.java:77]: This method is only called when the user does not set the index name and a name is generated. I don't think it is a good idea to generate a name that causes an error, that's why I truncate the name. I think we should keep it.

File: langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/JSONIndexBuilder.java
Comment (fmeheust) [reply langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/JSONIndexBuilder.java:98]: I don't understand where you would like me to add the validations. The method key(String, Class<?>, Order) is used to index a JSON key of the metadata column. If any of the parameters passed is null or empty (for the key), an exception will be thrown by the ensureXXX method.
```java
Index.jsonIndexBuilder()
        .createOption(CreateOption.CREATE_OR_REPLACE)
        .key("name", String.class, JSONIndexBuilder.Order.ASC)
        .key("year", Integer.class, JSONIndexBuilder.Order.DESC)
        .build()
```

File: langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/IndexBuilder.java
Comment (Michael-A-McMahon) [reply langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/IndexBuilder.java:77]: Oh, I see. Makes total sense!

File: langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/JSONIndexBuilder.java
Comment (Michael-A-McMahon) [reply langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/JSONIndexBuilder.java:98]: Sorry, I got confused. I thought we were in a build() method here. What you have is good!

**Pull request merged**
